SlevomatCodingStandard/Sniffs/ControlStructures/AbstractControlStructureSpacingSniff.php

```diff
@@ -0,0 +1,388 @@
+
+namespace SlevomatCodingStandard\Sniffs\ControlStructures;
+
+use Exception;
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Sniffs\Sniff;
+use PHP_CodeSniffer\Util\Tokens;
+use SlevomatCodingStandard\Helpers\SniffSettingsHelper;
+use SlevomatCodingStandard\Helpers\TokenHelper;
+use SlevomatCodingStandard\Sniffs\Namespaces\UndefinedKeywordTokenException;
+use function abs;
+use function array_key_exists;
+use function array_map;
+use function constant;
+use function count;
+use function defined;
+use function in_array;
+use function sprintf;
+use function strlen;
+use function substr;
+use function substr_count;
+use const T_ANON_CLASS;
+use const T_CASE;
+use const T_CATCH;
+use const T_CLOSE_CURLY_BRACKET;
+use const T_CLOSURE;
+use const T_COLON;
+use const T_COMMENT;
+use const T_DEFAULT;
+use const T_DO;
+use const T_ELSE;
+use const T_ELSEIF;
+use const T_FINALLY;
+use const T_FOR;
+use const T_FOREACH;
+use const T_IF;
+use const T_OPEN_CURLY_BRACKET;
+use const T_OPEN_SHORT_ARRAY;
+use const T_OPEN_TAG;
+use const T_SEMICOLON;
+use const T_SWITCH;
+use const T_TRY;
+use const T_WHILE;
+use const T_WHITESPACE;
+use const T_YIELD;
+use const T_YIELD_FROM;
+
+abstract class AbstractControlStructureSpacingSniff implements Sniff
+{
+
+	public const CODE_INCORRECT_LINES_COUNT_BEFORE_CONTROL_STRUCTURE = 'IncorrectLinesCountBeforeControlStructure';
+	public const CODE_INCORRECT_LINES_COUNT_BEFORE_FIRST_CONTROL_STRUCTURE = 'IncorrectLinesCountBeforeFirstControlStructure';
+	public const CODE_INCORRECT_LINES_COUNT_AFTER_CONTROL_STRUCTURE = 'IncorrectLinesCountAfterControlStructure';
+	public const CODE_INCORRECT_LINES_COUNT_AFTER_LAST_CONTROL_STRUCTURE = 'IncorrectLinesCountAfterLastControlStructure';
+
+	/** @var int */
+	public $linesCountBeforeControlStructure = 1;
+
+	/** @var int */
+	public $linesCountBeforeFirstControlStructure = 0;
+
+	/** @var int */
+	public $linesCountAfterControlStructure = 1;
+
+	/** @var int */
+	public $linesCountAfterLastControlStructure = 0;
+
+	/** @var bool */
+	public $allowSingleLineYieldStacking = true;
+
+	/** @var string[] */
+	public $tokensToCheck = [];
+
+	/** @var (string|int)[]|null */
+	private $normalizedTokensToCheck;
+
+	/**
+	 * @return (int|string)[]
+	 */
+	public function register(): array
+	{
+		return $this->getTokensToCheck();
+	}
+
+	/**
+	 * @phpcsSuppress SlevomatCodingStandard.TypeHints.TypeHintDeclaration.MissingParameterTypeHint
+	 * @param \PHP_CodeSniffer\Files\File $phpcsFile
+	 * @param int $controlStructurePointer
+	 */
+	public function process(File $phpcsFile, $controlStructurePointer): void
+	{
+		$this->checkLinesBefore($phpcsFile, $controlStructurePointer);
+		$this->checkLinesAfter($phpcsFile, $controlStructurePointer);
+	}
+
+	/**
+	 * @return int[]
+	 */
+	abstract protected function getDefaultTokensToCheck(): array;
+
+	/**
+	 * @return (int|string)[]
+	 */
+	private function getTokensToCheck(): array
+	{
+		if ($this->normalizedTokensToCheck === null) {
```

Comment:
We should use `getDefaultTokensToCheck()` to also filter only supported tokens. So maybe the name `getSupportedTokens` would be better.

Reply:
You mean to now allow other tokens then the default ones? That would break one test with `switch`, where custom tokens `T_CASE` and `T_DEFAULT` are used ...

Wouldn't it be better to leave it "open"? If it works for users to use different kinds of tokens, so be it...

Reply:
I don't want reports that something with custom tokens is not work as expected. 

So we need `getSupportedTokens` and also `getDefaultTokensToCheck()` because of `T_CASE` and `T_DEFAULT`.

Reply:
Oh ok, so `T_CASE` and `T_DEFAULT` should be supported but not checked by default ... got it

Reply:
It looks there another option :)

The support for `T_CASE` and `T_DEFAULT` was added later and are not in default tokens because of backward compatibility.

So we can have only `getSupportedTokens`. `T_CASE` and `T_DEFAULT` could be among supported tokens in `BlockControlStructureSpacingSniff` and not among in the old `ControlStructureSpacingSniff`

Reply:
Ah, I think I'm getting a bit lost here :D Check the latest commit and we will see... :D

Reply:
@grongor It's ok, I will merge it :)

Please squash and rebase.